src/AtlasAdmin.ts

```diff
@@ -0,0 +1,173 @@
+    const options = DEFAULT_ATLAS_API_OPTIONS;
+    options['method'] = 'GET';
+
+    try {
+      const { data, res } = await request(url, options);
+      if (res.statusCode !== 200) {
+        log.error(res);
+        throw res;
+      }
+
+      return data.find((index: any) => index.name === indexName);
+    } catch (e) {
+      log.error(`Error while fetching search index: ${JSON.stringify(e)}`);
+      throw e;
+    }
+  }
+
+  private async createSearchIndex(searchIndex: IndexMappings, taxonomy: Taxonomy) {
+    log.info('creating Atlas search index');
+    const url = `${this.baseUrl}`;
+    const options = this._getPostPatchOptions(searchIndex, taxonomy);
+    options['method'] = 'POST';
+
+    try {
+      const { res, data } = await request(url, options);
+      if (res.statusCode !== 200) {
+        log.error(res);
+        throw res;
+      }
+
+      return data;
+    } catch (e) {
+      log.error(`Error while creating new search index: ${JSON.stringify(e)}`);
+      throw e;
+    }
+  }
+
+  private async updateSearchindex(indexId: string, searchIndex: IndexMappings, taxonomy: Taxonomy) {
+    log.info('updating Atlas search index');
+    const url = `${this.baseUrl}/${indexId}`;
+
+    const options = this._getPostPatchOptions(searchIndex, taxonomy);
+    options['method'] = 'PATCH';
+
+    try {
+      const { res, data } = await request(url, options);
+      if (res.statusCode !== 200) {
+        log.error(res);
+        throw res;
+      }
+
+      return data;
+    } catch (e) {
+      log.error(`Error while updating search index: ${JSON.stringify(e)}`);
+      throw e;
+    }
+  }
+
+  private _getPostPatchOptions(searchIndex: IndexMappings, taxonomy: Taxonomy) {
+    const options = DEFAULT_ATLAS_API_OPTIONS;
+    options['data'] = this._insertTaxonomyIntoSearchIndex(searchIndex, taxonomy);
+    options['data'] = Object.assign(options['data'], {
+      collectionName: COLLECTION_NAME,
+      database: DB,
+      name: SEARCH_INDEX,
+    });
+    return options;
+  }
+
+  /**
+   * takes in Search Index config and parsed taxonomy toml
+   * inserts into Search Index and returns
+   *
+   * @param searchIndex
+   * @param taxonomy
+   * @returns Search Index with faceted indexes added from taxonomy
+   */
+  private _insertTaxonomyIntoSearchIndex(searchIndex: IndexMappings, taxonomy: Taxonomy) {
+    const keyList = getFacetKeys(taxonomy);
+    searchIndex['mappings']['fields']['facets'] = {
+      type: 'document',
+      fields: {},
+    };
+    for (const facetKey of keyList) {
+      searchIndex['mappings']['fields']['facets']['fields'][facetKey] = [
+        {
+          type: 'string',
+        },
+        {
+          type: 'stringFacet',
+        },
+      ];
+    }
+    return searchIndex;
+  }
+}
+
+// TODO: possibly move to taxonomy utils
+const getFacetKeys = (taxonomy: Taxonomy) => {
+  const keyList: string[] = [];
+  const pushKeys = (baseStr = '', currentRecord: Taxonomy) => {
```

Comment:
Super NIT: If `baseStr` is the second argument for the function, this would allow us to call `pushKeys` on line 171 without the redundant extra string.

Reply:
Done!